pkg/querier/querier.go

```diff
@@ -1144,6 +1145,62 @@ func (q *SingleTenantQuerier) DetectedFields(ctx context.Context, req *logproto.
 	}, nil
 }
 
+func (q *SingleTenantQuerier) StructuredMetadata(ctx context.Context, req *logproto.StructuredMetadataKeysRequest) (*logproto.StructuredMetadataKeysResponse, error) {
+	expr, err := syntax.ParseLogSelector(req.Query, true)
+	if err != nil {
+		return nil, err
+	}
+	params := logql.SelectLogParams{
+		QueryRequest: &logproto.QueryRequest{
+			Start:     req.Start,
+			End:       req.End,
+			Limit:     1000,
+			Direction: logproto.BACKWARD,
+			Selector:  expr.String(),
+			Plan: &plan.QueryPlan{
+				AST: expr,
+			},
+		},
+	}
+
+	iters, err := q.SelectLogs(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(twhitney): we can improve this in the future as we don't need to iterate lines for structured metadata, just chunks
+	streams, err := streamsForFieldDetection(iters, 1000)
+	if err != nil {
+		return nil, err
+	}
+
+	structuredMetadata := parseStructuredMetadata(ctx, streams)
+
+	return &logproto.StructuredMetadataKeysResponse{
+		Keys: structuredMetadata,
+	}, nil
+}
+
+func parseStructuredMetadata(ctx context.Context, streams logqlmodel.Streams) []string {
+	lbls := []string{}
+
+	for _, stream := range streams {
+		level.Debug(spanlogger.FromContext(ctx)).Log(
+			"structured_metadata", "true",
+			"msg", fmt.Sprintf("looking for structured metadata in stream %d with %d lines", stream.Hash, len(stream.Entries)))
+
+		for _, entry := range stream.Entries {
+			for _, l := range entry.StructuredMetadata {
+				if !slices.Contains(lbls, l.Name) {
```

Comment:
you're better of using a map to ensure unicity. Otherwise you doing a lot of iteration as the slices grows.

Once the map is full, you can then transform into an array and sort it for stability.

Reply:
map[string]struct{}{}

Reply:
Like https://github.com/grafana/loki/pull/13119/files#diff-a0b881d1b7b99f439716f189b50eef90ccdaea00845b087c0dbce0e148aa2c0eR1184-R1202 ?

Reply:
>Once the map is full, you can then transform into an array and sort it for stability.

do i have to sort it?

Reply:
In my experience anything coming out of a map key is better sorted because the order will be unstable and makes testing or client usage harder.